compiler/luci/export/src/CircleExporterImpl.cpp

```diff
@@ -222,19 +222,42 @@ void CircleExporterImpl::exportModule(Module *module)
   std::string description_str = "nnpackage";
   auto description = _builder.CreateString(description_str);
 
-  // create array of buffers
-  auto buffers = _builder.CreateVector(md._buffers);
-
-  // empty metadata
-  std::vector<int> metadata_buffer_vec;
-  auto metadata_buffer = _builder.CreateVector(metadata_buffer_vec);
+  // metadata
+  std::vector<flatbuffers::Offset<circle::Metadata>> metadata_vec;
+  {
+    // Insert origin node names
+    auto origin_name_buffer_id = static_cast<uint32_t>(md._buffers.size());
+    auto origin_name_data = std::vector<uint8_t>(); // TODO Fill data using md._metadata
+    auto origin_name_vec_offset =
+      _builder.CreateVector(origin_name_data.data(), origin_name_data.size());
+    auto origin_name_offset = CreateBuffer(_builder, origin_name_vec_offset);
+    md._buffers.push_back(origin_name_offset);
+    Offset<circle::Metadata> origin_name_metadata =
+      CreateMetadata(_builder, _builder.CreateString("origin_names_list"), origin_name_buffer_id);
+    metadata_vec.push_back(origin_name_metadata);
+
+    // Insert origin mapping data
+    auto origin_map_buffer_id = static_cast<uint32_t>(md._buffers.size());
+    auto origin_map_data = std::vector<uint8_t>(); // TODO Fill data using md._metadata
+    auto origin_map_vec_offset =
+      _builder.CreateVector(origin_map_data.data(), origin_map_data.size());
+    auto origin_map_offset = CreateBuffer(_builder, origin_map_vec_offset);
+    md._buffers.push_back(origin_map_offset);
+    Offset<circle::Metadata> origin_map_metadata =
+      CreateMetadata(_builder, _builder.CreateString("origin_mapping"), origin_map_buffer_id);
+    metadata_vec.push_back(origin_map_metadata);
+  }
+  auto metadata = _builder.CreateVector(std::vector<Offset<Metadata>>(metadata_vec));
 
   // This version is taken from comment in fbs
   constexpr uint32_t version = 0;
 
+  // create array of buffers
+  auto buffers = _builder.CreateVector(md._buffers);
```

Comment:
metadata and buffers  creation order has changed. is there any particular reason for this?

Reply:
@seanshpark I guess it is to add `Metadata` to buffers.
```flatbuffers
table Metadata {
  // A human readable string to uniquely identify a Metadata.
  name:string;
  // An index to the buffers table.
  buffer:uint;
}
```
As you can see, `Metadata` stores its data to buffer.

Reply:
@mhs4670go , isn't this `buffers` for Tensors? below line `buffers, metadata_buffer`  shows they are distinct items.
or maybe I understand incorrectly...

Reply:
https://github.com/tensorflow/tensorflow/issues/30387#issuecomment-508681200
> We don't have standardized methods to read from or write to it.

```flatbuffers
table Metadata {
  // A human readable string to uniquely identify a Metadata.
  name:string;
  // An index to the buffers table.
  buffer:uint;
}
```

I couldn't find any document about where the buffer would be used except comments in schema. As comment said, `Metadata` has `buffer` field that points to buffers table.

Reply:
From `schema.fbs`

```
  // Buffers of the model.
  // Note the 0th entry of this array must be an empty buffer (sentinel).
  // This is a convention so that tensors without a buffer can provide 0 as
  // their buffer.
  buffers:[Buffer];

  // Metadata about the model.  Indirects into the existings buffers list.
  metadata_buffer:[int];
```

Reply:
> metadata and buffers creation order has changed. is there any particular reason for this?

This is because metadata should be included in the buffer too.

>  isn't this buffers for Tensors? below line buffers, metadata_buffer shows they are distinct items.
or maybe I understand incorrectly...

Each items are distinct, but there is only one buffer table.
In `circle_schema.fbs`, following items is included.
```
table Model {
  ...

  // Buffers of the model.
  // Note the 0th entry of this array must be an empty buffer (sentinel).
  // This is a convention so that tensors without a buffer can provide 0 as
  // their buffer.
  buffers:[Buffer];

  // Metadata about the model. Indirects into the existings buffers list.
  // Deprecated, prefer to use metadata field.
  metadata_buffer:[int];

  // Metadata about the model.
  metadata:[Metadata];
}

root_type Model;
```
`Model` has only one buffer table(`buffers`) and all the buffer indices in the `Model` should point out one of the indices.
Following example might be helpful.
```
Model
{
  buffers
  {
    Buffer 0 : Empty
    Buffer 1 : Data for Tensor1
    Buffer 2 : Data for Tensor2
    Buffer 3 : Data for metadata1
    Buffer 4 : Data for metadata2
  }

  metadata
  {
    {
      name : "Metadata1"
      buffer : 3
    }
    {
      name : "Metadata2"
      buffer : 4
    }
  }

  subgraphs
  {
    tensors
    {
      Tensor 0 : Do not use buffer
      Tensor 1 : Use buffer 1
      Tensor 2 : Use buffer 2
    }
  }
}
```